### PR TITLE
Eng 12877 nibbledelete sysproc

### PR DESCRIPTION
**Pull request by luyangco**
The sysproc copies pattern from windowing app, it has three statements:

select numbers of rows that meet the delete criteria, find the value sits on boundary of delete chunk size, if rows to be deleted are more than chunk size, delete rows based on the value given in parameter, or the boundary value.

### CODE REVIEW

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        Procedure newCatProc = StatementCompiler.compileNibbleDeleteProcedure(catTable,
+                this.getClass().getName() + "-" + tableName, column, comparisonConstant);
+
+
+        Statement countStmt = newCatProc.getStatements().get(VoltDB.ANON_STMT_NAME + "0");
+        Statement deleteStmt = newCatProc.getStatements().get(VoltDB.ANON_STMT_NAME + "1");
+        Statement valueAtStmt = newCatProc.getStatements().get(VoltDB.ANON_STMT_NAME + "2");
+        if (countStmt == null || deleteStmt == null || valueAtStmt == null) {
+            throw new VoltAbortException(
+                    String.format("Unable to find SQL statement for found table %s: BAD",
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java]: This is not possible to happen, right?

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    /**
+     * Execute a set of queued inserts. Ensure each insert successfully
+     * inserts one row. Throw exception if not.
+     *
+     * @return Count of rows inserted or upserted.
+     * @throws VoltAbortException if any failure at all.
+     */
+    VoltTable executeSQLOnTheFly(Statement catStmt, Object[] params) throws VoltAbortException {
+        // Create a SQLStmt instance on the fly
+        // This unusual to do, as they are typically required to be final instance variables.
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java]: This is unusual to do?

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                            actualType.toString(), expectedType.toString()));
+        }
+
+        if (comparison < ComparisonConstant.GREATER_THAN.ordinal() ||
+                comparison > ComparisonConstant.EQUAL.ordinal()) {
+            throw new VoltAbortException("Invalid comparison constant: " + comparison);
+        }
+        ComparisonConstant comparisonConstant = ComparisonConstant.values()[comparison];
+
+        // TODO: should cache the plan in somewhere
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java]: Yes, every invocation will trigger the planning and the planning function is synchronized. 

Also, does it mean we create a Procedure for every invocation as well? Anything we need to worry about?

File: src/frontend/org/voltdb/compiler/StatementCompiler.java [hunk 4]
```diff
@@ -622,4 +625,155 @@ public static Procedure compileDefaultProcedure(PlannerTool plannerTool, Procedu
...
+            Column col, ComparisonConstant comp) {
+        Procedure newCatProc = addProcedure(catTable, procName);
+
+        String countingQuery = genSelectSql(catTable, col, comp);
+        addStatement(catTable, newCatProc, countingQuery, "0");
+
+        String deleteQuery = genDeleteSql(catTable, col, comp);
+        addStatement(catTable, newCatProc, deleteQuery, "1");
+
+        String valueAtQuery = genValueAtOffsetSql(catTable, col);
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/compiler/StatementCompiler.java]: So this is not the ben's midpoint guessing strategy. I am ok as the initial version.

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteMP.java [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+    public DependencyPair executePlanFragment(
+            Map<Integer, List<VoltTable>> dependencies, long fragmentId,
+            ParameterSet params, SystemProcedureExecutionContext context) {
+        // Never called, we do all the work in run()
+        return null;
+    }
+
+    /**
+     * Execute a set of queued inserts. Ensure each insert successfully
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/sysprocs/NibbleDeleteMP.java]: Similar comments as the NibbleDeleteSp.

**Comment by xinjiacs** [comment]
Although these two sysprocs are not intended for external usage, we may put these logics directly into the Hugg's NT procedure @NibbleDelete, and use @AdHoc system proc directly. In this way, it handles the query planning and cache issue, making it a simplified and flexible solution. 

Importantly, we are trying to compute the next batch timestamp point for delete query, it does not need to be accurate for the middle value. For each partition, the NT procedure will do 1) select min(timestamp), count(*)... , 2) guessing a point value 3) run delete query with that point value. 

Maybe we should wait for @jhugg , he may have more thoughts on the design.

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+        VoltType expectedType = VoltType.values()[column.getType()];
+        VoltType actualType = VoltType.typeFromObject(params[0]);
+        if (actualType != expectedType) {
+            throw new VoltAbortException(
+                    String.format("Parameter type %s doesn't match column type %s",
+                            actualType.toString(), expectedType.toString()));
+        }
+
+        if (comparison < ComparisonConstant.GREATER_THAN.ordinal() ||
```
[the remark this answers is not in the document]
Comment (luyangco) [reply src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java]: Sure.

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java (hunk 3 above)
Comment (luyangco) [reply src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java]: NibbleDelete can be treated like default procedure. Procedure is cached at defaultProcCache in each site.

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java (hunk 1 above)
Comment (luyangco) [reply src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java]: I don't think it's possible to fail here, but just a bullet-proof vest.

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+    public DependencyPair executePlanFragment(
+            Map<Integer, List<VoltTable>> dependencies, long fragmentId,
+            ParameterSet params, SystemProcedureExecutionContext context) {
+        // Never called, we do all the work in run()
+        return null;
+    }
+
+    /**
+     * Execute a set of queued inserts. Ensure each insert successfully
```
[the remark this answers is not in the document]
Comment (luyangco) [reply src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java]: Will change it.

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteMP.java (hunk 5 above)
Comment (luyangco) [reply src/frontend/org/voltdb/sysprocs/NibbleDeleteMP.java]: Will change it.

File: src/frontend/org/voltdb/compiler/StatementCompiler.java (hunk 4 above)
Comment (luyangco) [reply src/frontend/org/voltdb/compiler/StatementCompiler.java]: If we solve the performance issue in query
```sql 
select * from t order by ts offset N limit 1
```
Then we don't need to use midpoint guessing strategy.

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+     * @param table       value to compare
+     * @param chunksize   maximum number of rows allow to be deleted
+     * @return how many rows are deleted and how many rows left to be deleted (if any)
+     */
+    public VoltTable run(SystemProcedureExecutionContext ctx, int partitionParam, String tableName, String columnName,
+            int comparison, VoltTable table, long chunksize) {
+        // Some basic checks
+        Table catTable = ctx.getDatabase().getTables().getIgnoreCase(tableName);
+        if (catTable == null) {
+            throw new VoltAbortException("Table not present in catalog");
```
[the remark this answers is not in the document]
Comment (luyangco) [reply src/frontend/org/voltdb/sysprocs/NibbleDeleteSP.java]: Will do.

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteBase.java [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+        voltQueueSQL(stmt, params);
+        return voltExecuteSQL()[0];
+    }
+
+    VoltTable nibbleDeleteCommon(SystemProcedureExecutionContext ctx,
+                                 String tableName,
+                                 String columnName,
+                                 int comparison,
+                                 VoltTable table,
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/sysprocs/NibbleDeleteBase.java]: Rename "table" to "paramTable"?

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteBase.java [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        // so far should only be single column, single row table
+        int columnCount = table.getColumnCount();
+        if (columnCount > 1) {
+            throw new VoltAbortException("More than one input parameter is not supported right now.");
+        }
+        assert(columnCount == 1);
+        table.resetRowPosition();
+        table.advanceRow();
+        Object[] params = new Object[columnCount];
+        for (int i = 0; i < columnCount; i++) {
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/sysprocs/NibbleDeleteBase.java]: I am not sure how much value left to keep "columnCount" other than assuming 1. The other places does assume that already. 

This API already works for the simple use case and TTL later, maybe changing the parameter voltdb to be an java object as well for simplicity. The harder AdHoc Delete case will definitely need other solution.

File: src/frontend/org/voltdb/sysprocs/NibbleDeleteBase.java [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        result = executePrecompiledSQL(countStmt, params, replicated);
+        long rowCount = result.asScalarLong();
+        // If number of rows meet the criteria is more than chunk size, pick the column value
+        // which offset equals to chunk size as new predicate.
+        // TODO:how to delete rows match exactly to the chunk size while not causing the
+        // non-deterministic issue?
+        // Please be noted that it means rows be deleted can be more than chunk size, normally
+        // data has higher cardinality won't exceed the limit a lot,  but low cardinality data
+        // might cause this procedure to delete a large number of rows than the limit chunk size.
+        if (op != ComparisonConstant.EQUAL) {
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/sysprocs/NibbleDeleteBase.java]: Also, if we can assume the length of "params" is 1, we can simply the following delete stmt as:
```java
        Object paramValue = params[0];
        if (op == ComparisonConstant.EQUAL || rowCount <= chunksize) {
            cutoffValue = paramValue;
        } else {
            Object[] newParams = new Object[]{paramValue, chunksize - 1};

            result = executePrecompiledSQL(valueAtStmt, newParams, replicated);
            cutoffValue = result.fetchRow(0).get(0, actualType);
        }
        result = executePrecompiledSQL(deleteStmt, new Object[] {cutoffValue}, replicated);
```

File: src/frontend/org/voltdb/compiler/StatementCompiler.java [hunk 12]
```diff
@@ -622,4 +625,169 @@ public static Procedure compileDefaultProcedure(PlannerTool plannerTool, Procedu
         fragment.setPlannodetree(bin64String);
         return jsonBytes;
     }
+
+    private static String genSelectSql(Table table, Column column,
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/compiler/StatementCompiler.java]: As these functions are not intended for general use, maybe add "NibbleDelete" to these function names for clarification?

File: tests/frontend/org/voltdb/sysprocs/TestNibbleDelete.java [hunk 13]
```diff
@@ -0,0 +1,430 @@
...
+        }
+
+        pair = new Pair<>(0l, Long.MAX_VALUE); // <deleted, toBeDeleted>
+        loop = 0;
+        existingRows = numberOfItems;
+        while (pair.getSecond() > 0) {
+            pair = nibbleDeleteReplicated(op, ts, existingRows);
+            existingRows -= pair.getFirst();
+            if (++loop > 100) {
+                fail("Make no progress on delete, something wrong happens in @NibbleDeleteSP");
```
Comment (xinjiacs) [inline tests/frontend/org/voltdb/sysprocs/TestNibbleDelete.java]: NibbleDeleteMP?

File: tests/frontend/org/voltdb/sysprocs/TestNibbleDelete.java [hunk 14]
```diff
@@ -0,0 +1,430 @@
...
+                    table,
+                    500);
+            VoltTable result = response.getResults()[0];
+            assertEquals(1, result.getRowCount());
+            result.advanceRow();
+            deleted = result.getLong("DELETED_ROWS");
+            long leftoverRows = result.getLong("LEFTOVER_ROWS");
+            toBeDeleted += leftoverRows;
+        } catch (ProcCallException e) {
+            fail("Fail to run NibbleDeleteSP: " + e.getMessage());
```
Comment (xinjiacs) [inline tests/frontend/org/voltdb/sysprocs/TestNibbleDelete.java]: NibbleDeleteMP?

File: src/frontend/org/voltdb/compiler/StatementCompiler.java [hunk 15]
```diff
@@ -622,4 +625,169 @@ public static Procedure compileDefaultProcedure(PlannerTool plannerTool, Procedu
...
+        sb.append(" WHERE " + column.getName() + " " + comparison.toString() + " ?;");
+        return sb.toString();
+    }
+
+    private static String genValueAtOffsetSql(Table table, Column column,
+            ComparisonConstant comparison) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("SELECT " + column.getName() + " FROM " + table.getTypeName());
+        sb.append(" WHERE " + column.getName() + " " + comparison.toString() + " ?");
+        sb.append(" ORDER BY " + column.getName());
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/compiler/StatementCompiler.java]: Does this still mean doing an index scan anyway?

File: tests/frontend/org/voltdb/sysprocs/TestNibbleDelete.java [hunk 16]
```diff
@@ -0,0 +1,430 @@
...
+    {
+        System.out.println("testSelectByOffsetOnIndexColumn");
+
+        int numberOfItems = 100000;
+        VoltTable part_table = createTable(numberOfItems, 0, 0, 0, 0);
+        loadTable(m_client, "part", false, part_table);
+
+        String query = "select * from part order by ts offset 1000 limit 1";
+        CatalogContext context = VoltDB.instance().getCatalogContext();
+        PlannerTool plannerTool = context.m_ptool;
```
Comment (xinjiacs) [inline tests/frontend/org/voltdb/sysprocs/TestNibbleDelete.java]: query and plannerTool not used?

**Pull request merged**